### DISCUSSION TITLE
read roughness from pincipled bsdf

### DISCUSSION THREAD

**Discussion by zichenwang01** [mitsuba-renderer/mitsuba3#1152] [Q&A] [answered]
Hi,
I want to read specifically the roughness value from Mitsuba's principled bsdf within my custom renderer. I see there is a build-in function for reflectance
```
albedo = bsdf.eval_diffuse_reflectance(itx)
```

I wonder how can I do it for roughness? Thanks!

**Comment by njroussel** [comment]
Hi @zichenwang01 

You should  be able to do `bsdf.eval_attribute_1('roughness', si active)`.
The `eval_attribute_N` methods can evaluate any attribute/parameter that you would typically get when using `mi.traverse()`. The `N` serves as an indicator for the number of dimensions of the desired output value.

**Reply by zichenwang01** [reply]
Thanks for this very helpful information! I have a bit of trouble calling this function inside Python--
```
AttributeError: 'mitsuba.cuda_ad_rgb.BSDFPtr' object has no attribute 'eval_attribute_1'
```
where my code looks like
```
itx : SurfaceInteraction3f = scene.ray_intersect(ray)
bsdf : mi.BSDF = itx.bsdf()
roughness = bsdf.eval_attribute_1('roughness', itx)
```

**Reply by njroussel** [reply, marked as answer]
Hi @zichenwang01 

Sorry for the delay. Have you figured this out?
I just gave it another go, and it seems to run without any issues:

```python
import mitsuba as mi
import drjit as dr

mi.set_variant('cuda_ad_rgb')

scene = mi.load_dict(mi.cornell_box())
ray = mi.Ray3f(o=[0, 0, 3], d=[0, 0, -1])
si = scene.ray_intersect(ray)
bsdf = si.bsdf()
roughness = bsdf.eval_attribute_1('roughness', si)
print(f"{roughness=}") # return 0 but at least there aren't any errors
```

**Reply by jijiangli-swu-worker** [reply]
Hi, I was also confusing in this, so I may need for some help. For diffuse reflectence, put it onto the image was simple in the following code:
```python
            block = sensor.film().create_block()
            block.set_coalesce(block.coalesce() and spp >= 4)
            diffuse_alpha = dr.select(active, mi.Float(1), mi.Float(0))
            block.put(pos, ray.wavelengths, diffuse, diffuse_alpha)
```
This was like you do in `Integrator` render function. But for single channel like `roughness` or `depth`, could there be some way to put it on to a imageblock and output? Any response would be helpful. I try simple put single channel var into RGB image block, but I was scared about the value could be changed automaticlly when I extend one channel attribute into RGB image block. The same discusstion I have created https://github.com/mitsuba-renderer/mitsuba3/discussions/1394
Any response would be helpful!
